netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -117,14 +132,28 @@ public static FromChannelCredentialsResult from(ChannelCredentials creds) {
             new ByteArrayInputStream(tlsCreds.getPrivateKey()),
             tlsCreds.getPrivateKeyPassword());
       }
-      if (tlsCreds.getTrustManagers() != null) {
-        builder.trustManager(new FixedTrustManagerFactory(tlsCreds.getTrustManagers()));
-      } else if (tlsCreds.getRootCertificates() != null) {
-        builder.trustManager(new ByteArrayInputStream(tlsCreds.getRootCertificates()));
-      } // else use system default
+      Optional<TrustManager> x509ExtendedTrustManager;
       try {
-        return FromChannelCredentialsResult.negotiator(tlsClientFactory(builder.build()));
-      } catch (SSLException ex) {
+        if (tlsCreds.getTrustManagers() != null) {
+          builder.trustManager(new FixedTrustManagerFactory(tlsCreds.getTrustManagers()));
+          x509ExtendedTrustManager = tlsCreds.getTrustManagers().stream().filter(
+              trustManager -> trustManager instanceof X509ExtendedTrustManager).findFirst();
+        } else if (tlsCreds.getRootCertificates() != null) {
+          builder.trustManager(new ByteArrayInputStream(tlsCreds.getRootCertificates()));
```

Comment:
I'm very wary of assuming we're using the same implementation as the `builder`. I suggest have all these cases produce `TrustManager[] tms`, then after the if-elses call `builder.trustManager(tms)` and filter `tms` for x509ExtendedTrustManager.

Reply:
I didn't understand this comment.  When you say assuming same implementation as the builder are you referring to this line

builder.trustManager(new ByteArrayInputStream(tlsCreds.getRootCertificates()));

since the other 2 lines don't reference the builder. This is an existing statement. Implementation of TrustManager as the builder?

Reply:
`builder.trustManager()` is having the builder use some certs for its trust. But then you use a _different_ trust manager implementation for the per-RPC verification. I'm wary of them being different. There could be a performance reason to do so; I've not checked recently if tcnative has a fast path for simple file-based verification (to use boringssl/openssl instead of calling back into Java). But in general, creating two different instances (the code you added, and the existing builder.trustManager()) and hoping they behave the same is worrisome.

Reply:
Done.